contracts/forwarder/Forwarder.sol

```diff
@@ -59,16 +59,25 @@ contract Forwarder is IForwarder {
         _verifyAndUpdateNonce(req);
 
         bytes memory callData = abi.encodePacked(req.data, req.from);
-        require( gasleft()*63/64 >= req.gas, "FWD: insufficient gas" );
+        uint256 preBalance = address(this).balance;
+        require(preBalance >= req.value, "FWD: insufficient value");
+        require(gasleft()*63/64 >= req.gas, "FWD: insufficient gas" );
         require(req.validUntil == 0 || req.validUntil > block.number, "FWD: request expired");
-
         // solhint-disable-next-line avoid-low-level-calls
-        (success,ret) = req.to.call{gas : req.gas, value : req.value}(callData);
-        if ( address(this).balance>0 ) {
+        (success,ret) = req.to.call{gas: req.gas, value: req.value}(callData);
+
+        int256 excessBalance = int256(address(this).balance - preBalance + req.value);
```

Comment:
By the way, you don't check if the call succeeded or not. If the recipient reverted, the entire `req.value` will be stuck in Forwarder, won't it?